kotlin/export-tiles/src/main/java/com/example/exporttiles/MainActivity.kt

```diff
@@ -0,0 +1,240 @@
+    setContentView(R.layout.activity_main)
+
+    // create an ArcGISTiledLayer to use as the basemap
+    val tiledLayer = ArcGISTiledLayer(getString(R.string.world_street_map))
+    val map = ArcGISMap().apply {
+      basemap = Basemap(tiledLayer)
+      minScale = 10000000.0
+    }
+    // create a graphics overlay
+    val graphicsOverlay = GraphicsOverlay()
+    mapView.apply {
+      // set the map to the map view
+      this.map = map
+      setViewpoint(Viewpoint(51.5, 0.0, 10000000.0))
+      // add the graphics overlay to the map view
+      graphicsOverlays.add(graphicsOverlay)
+    }
+
+    // create a graphic to show a red box around the tiles we want to download
+    val downloadArea = Graphic()
+    graphicsOverlay.graphics.add(downloadArea)
+    val simpleLineSymbol = SimpleLineSymbol(SimpleLineSymbol.Style.SOLID, Color.RED, 2f)
+    downloadArea.symbol = simpleLineSymbol
+
+    // update the box whenever the viewpoint changes
+    mapView.addViewpointChangedListener {
+      if (mapView.map.loadStatus == LoadStatus.LOADED) {
+        // upper left corner of the downloaded tile cache area
+        val minScreenPoint = Point(150, 175)
+        // lower right corner of the downloaded tile cache area
+        val maxScreenPoint = Point(
+          mapView.getWidth() - 150,
+          mapView.getHeight() - 250
+        )
+        // convert screen points to map points
+        val minPoint: com.esri.arcgisruntime.geometry.Point? =
+          mapView?.screenToLocation(minScreenPoint)
+        val maxPoint: com.esri.arcgisruntime.geometry.Point? =
+          mapView?.screenToLocation(maxScreenPoint)
+        // use the points to define and return an envelope
+        val envelope = Envelope(minPoint, maxPoint)
+        downloadArea.geometry = envelope
+      }
+    }
+
+    // create up a temporary directory in the app's cache for saving exported tiles
+    val exportTilesDirectory = File(cacheDir, getString(R.string.tile_cache_folder))
+
+    // when the button is clicked, export the tiles to a temporary file
+    exportTilesButton.setOnClickListener {
+      val exportTileCacheTask = ExportTileCacheTask(tiledLayer.uri)
+      val parametersFuture: ListenableFuture<ExportTileCacheParameters> =
+        exportTileCacheTask.createDefaultExportTileCacheParametersAsync(
+          downloadArea.geometry,
+          mapView.mapScale,
+          tiledLayer.maxScale
+        )
+
+      parametersFuture.addDoneListener {
+        val parameters: ExportTileCacheParameters = parametersFuture.get()
+        // export tiles to temporary cache on device
+        val exportTileCacheJob =
+          exportTileCacheTask.exportTileCache(parameters, exportTilesDirectory.path + "file.tpk")
+        exportTileCacheJob.start()
+        // create a progress dialog to show export tile progress
+        val dialog = createProgressDialog(exportTileCacheJob).apply {
+          getButton(DialogInterface.BUTTON_NEUTRAL).setTextColor(R.style.progressBarBlue)
```

Comment:
Hmm this line gives a null object reference for me when I hit "Export Tiles"

Reply:
it's started doing that for me too: I'll just remove it, it's not important.